tests/integration/components/bs-form-element-test.js

```diff
@@ -79,4 +94,53 @@ module('Integration | Component | bs form element', function(hooks) {
     await triggerEvent('form', 'submit');
     assert.ok(find('.form-group').classList.contains('has-error'), 'form element group has error class');
   });
+
+
+  test('more complicated validations', async function(assert) {
+    let model = EmberObject.create({
+      name: '',
+      password: null,
+      passwordConfirmation: null,
+      email: ''
+    });
+
+    this.set('model', model);
+    this.set('validation', extendedValidation);
+    this.actions.submitAction = function() {
+      assert.ok(false, 'submit action must not been called.');
+    };
+    this.actions.invalidAction = function() {
+      assert.ok(true, 'Invalid action has been called.');
+    };
+
+    this.actions.validate = function() {
+      return 'invalid';
+    };
+
+    await render(hbs`
+      {{#bs-form model=(changeset model validation) onSubmit=(action "submitAction") onInvalid=(action "invalidAction") as |form|}}
+        {{form.element id="name" label="Name" property="name"}}
+        {{form.element id="email" label="Email" property="email"}}
+        {{form.element id="password" label="Password" property="password"}}
+        {{form.element id="password-confirmation" label="Password confirmation" property="passwordConfirmation"}}
+      {{/bs-form}}
+    `);
+
+    assert.expect(6);
+
+    await typeIn('#password-field', 'bad');
```

Comment:
Any specific reason to use `typeIn` here instead of `fillIn`? Just wondering, does `typeIn` uncover things that `fillIn` wouldn't, as it seems to better match what a real user would do?

Reply:
No specific reason.  I will switch it and rebase.